File: springboot/src/main/java/com/mcp/server/service/WeatherService.java
package com.mcp.server.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Mono;

import java.util.Map;

@Service
public class WeatherService {

    private static final String BASE_URL = "https://api.openweathermap.org/data/2.5/weather";
    
    @Value("${mcp.api-key.openweathermap}")
    private String apiKey;
    
    private final WebClient.Builder webClientBuilder;

    public WeatherService(WebClient.Builder webClientBuilder) {
        this.webClientBuilder = webClientBuilder;
    }

    public Mono<Map<String, Object>> getWeatherData(String city) {
        return webClientBuilder.build()
            .get()
            .uri(uriBuilder -> uriBuilder
                .scheme("https")
                .host("api.openweathermap.org")
                .path("/data/2.5/weather")
                .queryParam("q", city)
                .queryParam("appid", apiKey)
                .queryParam("units", "metric")
                .build())
            .retrieve()
            .bodyToMono(Map.class)
            .map(map -> (Map<String, Object>) map)
            .onErrorMap(e -> new RuntimeException("Failed to fetch weather data: " + e.getMessage(), e));
    }

    public Mono<Map<String, Object>> getAirQualityData(double lat, double lon) {
        return webClientBuilder.build()
            .get()
            .uri(uriBuilder -> uriBuilder
                .scheme("http")
                .host("api.openweathermap.org")
                .path("/data/2.5/air_pollution")
                .queryParam("lat", lat)
                .queryParam("lon", lon)
                .queryParam("appid", apiKey)
                .build())
            .retrieve()
            .bodyToMono(Map.class)
            .map(map -> (Map<String, Object>) map)
            .onErrorMap(e -> new RuntimeException("Failed to fetch air quality data: " + e.getMessage(), e));
    }
}

File: python/src/services/weather_service.py
import httpx
from typing import Dict, Any, Optional
from src.config import settings

class WeatherService:
    BASE_URL = "https://api.openweathermap.org/data/2.5/weather"

    async def get_weather_data(self, city: str) -> Dict[str, Any]:
        if not settings.OPENWEATHERMAP_API_KEY:
             raise ValueError("OpenWeatherMap API Key is missing")

        params = {
            "q": city,
            "appid": settings.OPENWEATHERMAP_API_KEY,
            "units": "metric" # Celsius
        }

        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(self.BASE_URL, params=params)
                response.raise_for_status()
                return response.json()
            except httpx.HTTPStatusError as e:
                if e.response.status_code == 404:
                    raise ValueError(f"City not found: {city}")
                raise Exception(f"OpenWeatherMap API error: {e.response.status_code}")
            except Exception as e:
                raise Exception(f"Failed to fetch weather data: {str(e)}")

    async def get_air_quality_data(self, lat: float, lon: float) -> Dict[str, Any]:
        if not settings.OPENWEATHERMAP_API_KEY:
             raise ValueError("OpenWeatherMap API Key is missing")

        url = "http://api.openweathermap.org/data/2.5/air_pollution"
        params = {
            "lat": lat,
            "lon": lon,
            "appid": settings.OPENWEATHERMAP_API_KEY
        }

        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(url, params=params)
                response.raise_for_status()
                return response.json()
            except Exception as e:
                raise Exception(f"Failed to fetch air quality data: {str(e)}")

weather_service = WeatherService()
